privacyidea/api/lib/postpolicy.py

```diff
@@ -318,6 +318,54 @@ def no_detail_on_success(request, response):
     return response
 
 
+def preferred_client_mode(request, response):
+    """
+    This policy function is used to add the preferred client mode.
+    The admin can set the list of client modes in the policy in the
+    same order as  he preferred them. The faction will pick the first
+    client mode from the list, that is also in the multichallege and
+    set it as preferred client mode
+
+    :param request:
+    :param response:
+    :return:
+    """
+    content = response.json
+
+    # get the preferred client mode from a policy definition
+    detail_pol = Match.action_only(g, scope=SCOPE.AUTH, action=ACTION.PREFERREDCLIENTMODE)\
+        .action_values(unique=True)
+
+    if detail_pol:
+        preferred_client_mode_list = detail_pol.get("test")[0].split(", ")
+    else:
+        preferred_client_mode_list = ['poll', 'webauthn', 'interactive', 'u2f']
+    if content.get("detail"):
+        detail = content.get("detail")
+        if detail.get("multi_challenge"):
+            multi_challenge = detail.get("multi_challenge")
+            l = []
+            for x in multi_challenge:
+                l.append(x.get("client_mode"))
+            l=list(dict.fromkeys(l))
+
+            if preferred_client_mode_list[0] in l:
+                preferred = preferred_client_mode_list[0]
+                content.setdefault("detail", {})["preferred_client_mode"] = preferred
+            elif preferred_client_mode_list[1] in l:
+                preferred = preferred_client_mode_list[1]
+                content.setdefault("detail", {})["preferred_client_mode"] = preferred
+            elif preferred_client_mode_list[2] in l:
+                preferred = preferred_client_mode_list[2]
+                content.setdefault("detail", {})["preferred_client_mode"] = preferred
+            elif preferred_client_mode_list[3] in l:
+                preferred = preferred_client_mode_list[3]
+                content.setdefault("detail", {})["preferred_client_mode"] = preferred
```

Comment:
You could move the four `content.setdefault`s out of the if branch to only one call at the end.

Reply:
You could simplify getting the preferred value like this:

~~~~Python
preferred = [x for x in preferrend_client_mode_list if x in l][0]
~~~~
This would replace the lines 352 to 362.

Reply:
[34ee918](https://github.com/privacyidea/privacyidea/pull/3361/commits/34ee918173549cdef4dfeb1e000951457fff1cab)